File: backend/app/schemas/assistant.py
"""Contrats structurés de l'Assistant ComptaFlow."""

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field


class AssistantQuery(BaseModel):
    question: str = Field(min_length=2, max_length=1500)
    conversation_id: uuid.UUID | None = None
    context_document_id: uuid.UUID | None = None
    entreprise_id: uuid.UUID | None = None
    page: int = Field(default=1, ge=1, le=100)


class AssistantAction(BaseModel):
    label: str
    kind: Literal["route", "secure_file"]
    route: str | None = None
    api_path: str | None = None
    filename: str | None = None


class AssistantSource(BaseModel):
    resource_type: str
    resource_id: uuid.UUID
    label: str
    route: str | None = None


class AssistantDocumentCard(BaseModel):
    document_id: uuid.UUID
    entreprise_id: uuid.UUID | None = None
    entreprise: str | None = None
    categorie: str | None = None
    tiers: str | None = None
    numero_facture: str | None = None
    date_facture: date | None = None
    date_importation: datetime
    montant_ht: Decimal | None = None
    montant_tva: Decimal | None = None
    montant_ttc: Decimal | None = None
    devise: str = "MAD"
    statut_document: str
    statut_ecriture: str | None = None
    statut_paiement: str | None = None
    montant_regle: Decimal | None = None
    restant_du: Decimal | None = None
    saisie_topaze: bool = False
    anomalies: list[str] = Field(default_factory=list)
    donnees_extraites: dict[str, Any] | None = None
    actions: list[AssistantAction] = Field(default_factory=list)


class AssistantPaymentCard(BaseModel):
    mouvement_id: uuid.UUID
    ecriture_id: uuid.UUID
    date_operation: date
    montant_mouvement: Decimal
    montant_affecte: Decimal
    restant_du: Decimal
    reference: str | None = None
    libelle: str
    statut: str
    actions: list[AssistantAction] = Field(default_factory=list)


class AssistantCompanyCard(BaseModel):
    entreprise_id: uuid.UUID
    nom: str
    ice: str | None = None
    identifiant_fiscal: str | None = None
    is_active: bool
    actions: list[AssistantAction] = Field(default_factory=list)


class AssistantDataCard(BaseModel):
    """Carte générique sûre pour les résultats non documentaires."""

    resource_type: str
    resource_id: uuid.UUID
    title: str
    subtitle: str | None = None
    fields: dict[str, Any] = Field(default_factory=dict)
    actions: list[AssistantAction] = Field(default_factory=list)


class AssistantResponse(BaseModel):
    conversation_id: uuid.UUID
    response_type: Literal[
        "single_document", "document_list", "aggregate", "accounting_detail",
        "company_list", "message", "clarification", "no_result", "error",
        "data_list",
    ]
    intent: str
    message: str
    filters: dict[str, Any] = Field(default_factory=dict)
    documents: list[AssistantDocumentCard] = Field(default_factory=list)
    companies: list[AssistantCompanyCard] = Field(default_factory=list)
    payments: list[AssistantPaymentCard] = Field(default_factory=list)
    data: list[AssistantDataCard] = Field(default_factory=list)
    total_count: int | None = None
    aggregate: dict[str, Any] | None = None
    sources: list[AssistantSource] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    clarification_question: str | None = None
    context_document_id: uuid.UUID | None = None
    suggestions: list[str] = Field(default_factory=list)
    plan_summary: dict[str, Any] = Field(default_factory=dict)
    page: int = 1
    page_size: int = 10
    has_more: bool = False

File: frontend/src/types/assistant.ts
export interface AssistantAction {
  label: string;
  kind: "route" | "secure_file";
  route: string | null;
  api_path: string | null;
  filename: string | null;
}

export interface AssistantDocumentCard {
  document_id: string;
  entreprise_id: string | null;
  entreprise: string | null;
  categorie: string | null;
  tiers: string | null;
  numero_facture: string | null;
  date_facture: string | null;
  date_importation: string;
  montant_ht: string | number | null;
  montant_tva: string | number | null;
  montant_ttc: string | number | null;
  devise: string;
  statut_document: string;
  statut_ecriture: string | null;
  statut_paiement: string | null;
  montant_regle: string | number | null;
  restant_du: string | number | null;
  saisie_topaze: boolean;
  anomalies: string[];
  donnees_extraites: Record<string, unknown> | null;
  actions: AssistantAction[];
}

export interface AssistantPaymentCard {
  mouvement_id: string;
  ecriture_id: string;
  date_operation: string;
  montant_mouvement: string | number;
  montant_affecte: string | number;
  restant_du: string | number;
  reference: string | null;
  libelle: string;
  statut: string;
  actions: AssistantAction[];
}

export interface AssistantCompanyCard {
  entreprise_id: string;
  nom: string;
  ice: string | null;
  identifiant_fiscal: string | null;
  is_active: boolean;
  actions: AssistantAction[];
}

export interface AssistantDataCard {
  resource_type: string;
  resource_id: string;
  title: string;
  subtitle: string | null;
  fields: Record<string, unknown>;
  actions: AssistantAction[];
}

export interface AssistantResponse {
  conversation_id: string;
  response_type: "single_document" | "document_list" | "company_list" | "data_list" | "message" | "aggregate" | "accounting_detail" | "clarification" | "no_result" | "error";
  intent: string;
  message: string;
  filters: Record<string, unknown>;
  documents: AssistantDocumentCard[];
  companies: AssistantCompanyCard[];
  payments: AssistantPaymentCard[];
  data: AssistantDataCard[];
  total_count: number | null;
  aggregate: Record<string, unknown> | null;
  sources: Array<{ resource_type: string; resource_id: string; label: string; route: string | null }>;
  warnings: string[];
  clarification_question: string | null;
  context_document_id: string | null;
  suggestions: string[];
  plan_summary: Record<string, unknown>;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface AssistantQuery {
  question: string;
  conversation_id?: string | null;
  context_document_id?: string | null;
  entreprise_id?: string | null;
  page?: number;
}
